experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -875,10 +843,23 @@ void HandshakeSpeculationPass::runDynamaticPass() {
   // to satisfy their type requirements.
   if (failed(addNonSpecOp()))
     return signalPassFailure();
-}
 
-std::unique_ptr<dynamatic::DynamaticPass>
-dynamatic::experimental::speculation::createHandshakeSpeculation(
-    const std::string &jsonPath, bool automatic) {
-  return std::make_unique<HandshakeSpeculationPass>(jsonPath, automatic);
+  // Quick fix: branchDiscardCondNonspec's operands must be the loop condition
+  // The real loop condition only turns out after the placement of speculative
+  // units (Speculator or save-commit unit may produce this)
+  handshake::FuncOp funcOp = specOp1->getParentOfType<handshake::FuncOp>();
+  for (auto branch : funcOp.getOps<handshake::SpeculatingBranchOp>()) {
+    if (branch->getAttr("specv1_branchDiscardCondNonSpec")) {
+      unsigned bb = getLogicBB(specOp1).value();
+      ConditionalBranchOp controlBranch = findControlBranch(funcOp, bb);
+      if (controlBranch == nullptr) {
+        specOp1->emitError()
+            << "Could not find backedge within speculation bb.\n";
+        return signalPassFailure();
+      }
+      auto conditionOperand = controlBranch.getConditionOperand();
+      branch->setOperand(0, conditionOperand);
+      branch->setOperand(1, conditionOperand);
+    }
+  }
```

Comment:
The operands of this SpeculatingBranch didn't follow this figure actually.
The loop condition is not necessarily produced by the speculator.
It only becomes definite after placing speculator and save-commit units (because the loop condition may be generated by either of them), so I'm doing this at the end of the pass.

<img width="632" height="794" alt="image" src="https://github.com/user-attachments/assets/689947ea-8de5-4ad5-98ed-2fa5342ebee1" />